subsys/bluetooth/mesh/dfu_cli.c

```diff
@@ -960,11 +961,11 @@ const struct bt_mesh_model_op _bt_mesh_dfu_cli_op[] = {
 	BT_MESH_MODEL_OP_END,
 };
 
-static int dfu_cli_init(struct bt_mesh_model *mod)
+static int dfu_cli_init(const struct bt_mesh_model *mod)
 {
-	struct bt_mesh_dfu_cli *cli = mod->user_data;
+	struct bt_mesh_dfu_cli *cli = *(mod->user_data);
 
-	if (mod->elem_idx != 0) {
+	if (*(mod->elem_idx) != 0) {
```

Comment:
This approach looks cool, but I feel it is very bug prone as we will always need to remember dereference `user_data`, `elem_idx`, `mod_idx` and `flags`. Users will have to remember this for `user_data` and (probably) `elem_idx` and `mod_idx`. And specifically here, if one misses `*`, compiler won't warn.

Reply:
I think the previous PR was good enough.

Reply:
https://github.com/zephyrproject-rtos/zephyr/pull/62432#discussion_r1370247990

Reply:
maybe rename mod_idx to mod_idx_ptr flags to flags_ptr and so on.

Reply:
BTW: I don't want to change things over and over again, without reaching a consensus on change request.

Reply:
> [#62432 (comment)](https://github.com/zephyrproject-rtos/zephyr/pull/62432#discussion_r1370247990)

I can't agree with @omkar3141 comment. The need in struct is clear IMO: to split compile time and run time data.

> maybe rename mod_idx to mod_idx_ptr flags to flags_ptr and so on.

But this still doesn't help to avoid issues like `*elem_idx_ptr != 0` vs `elem_idx_ptr != 0`.

> BTW: I don't want to change things over and over again, without reaching a consensus on change request.

Understand. But I didn't ask for changing this again. I was OK with previous change tbh.

Reply:
I think the naming of the new structure in the last PR was not very good, it would make people confused. I agree with omkar on this point.

Reply:
Call it `bt_mesh_model_rt_ctx`, `bt_mesh_model_runtime_ctx`?

Reply:
> This approach looks cool, but I feel it is very bug prone as we will always need to remember dereference `user_data`, `elem_idx`, `mod_idx` and `flags`.

Regarding the problem you mentioned, the way I can think of is to define a macro, for example: #define BT_MESH_MODEL_ELEM_IDX(_mod) *(_mod)->elem_idx and then use macros to replace all member accesses.

Reply:
@PavelVPV changed.

Reply:
@PavelVPV @alxelax Since this PR contains a lot of modifications, I don't want to have constant modification conflicts if I keep this PR open. So, please take a look if have free time, thanks.